### PR TITLE
[Backport release-23.11] stdenv: substituteStream: deprecate --replace in favor of --replace-{fail,warn,quiet}

### PR DESCRIPTION
**Pull request by misuzu**
## Description of changes
Backport the following changes:
https://github.com/NixOS/nixpkgs/pull/260776
https://github.com/NixOS/nixpkgs/pull/286526
https://github.com/NixOS/nixpkgs/pull/287111

This should allow backporting changes like this: https://github.com/NixOS/nixpkgs/pull/292041

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/stdenv/generic/setup.sh [hunk 1]
```diff
@@ -815,15 +815,27 @@ fi
...
     while (( "$#" )); do
+        local replace_mode="$1"
         case "$1" in
             --replace)
+                # deprecated 2023-11-22
+                # this will either get removed, or switch to the behaviour of --replace-fail in the future
+                if ! "$_substituteStream_has_warned_replace_deprecation"; then
+                    echo "substituteStream(): WARNING: '--replace' is deprecated, use --replace-{fail,warn,quiet}. ($description)" >&2
+                    _substituteStream_has_warned_replace_deprecation=true
+                fi
```
Comment (pbsds) [inline pkgs/stdenv/generic/setup.sh:835]: Warning about this deprecation on 23.11 version, where the behavior was technically not deprecated at release, i believe is unsound.

**Review by pbsds — COMMENTED** [review]
Functionality LGTM, but I'm not sure if we should backport the release note and deprecation warning. 

This is technically not a 23.11 _deprecation_, it is support for future behavior on unstable and 24.05, analogous to [`__future__`](https://docs.python.org/3/library/__future__.html) in python

**Comment by misuzu** [comment]
Tested by building https://github.com/NixOS/nixpkgs/pull/292041 on top of this change

**Comment by misuzu** [comment]
> Functionality LGTM, but I'm not sure if we should backport the release note and deprecation warning.

Removed. Should I remove the changes to the docs too?

**Comment by pbsds** [comment]
> Should I remove the changes to the docs too?

Don't think that is necessary. LGTM, Thanks!

**Pull request merged**